packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,624 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
+        self.points.splice(indexStart, action.pointsFromDrag.length)
+        self.setDragPoint(self.points.at(-1))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-drag-point') {
+        self.spliceDragPointIndex = null
+        self.setDragPoint((self.pathIsClosed) ? null : self.points.at(-1))
+        self.setClosePoint((self.pathIsClosed) ? null : self.points.at(0))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-close-point') {
+        self.splicePoints.length = 0
+        self.setClosePoint(null)
+        self.spliceReverse = false
+        self.spliceThroughBeginning = false
+        self.spliceActive = false
+        self.spliceClosePointIndex = null
+        self.drawActions.pop()
+      } else if (action.type == 'splice-append') {
+        let indexStart = self.splicePoints.length - action.pointsFromDrag.length
+        self.splicePoints.splice(indexStart, action.pointsFromDrag.length)
+        self.setDragPoint(self.splicePoints.at(-1))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-end') {
+        if (action.throughBeginning) {
+          let spliceIndex = Math.abs(action.pointIndexDrag - action.pointIndexClose)
+
+          self.splicePoints = self.splicePoints.concat(self.points.slice(spliceIndex, self.points.length - 1));
+
+          if (action.isReversed) {
+            self.splicePoints = self.splicePoints.reverse()
+          }
+
+          // we add 1 because we don't splice out either drag/close point
+          self.points.splice(spliceIndex + 1, self.points.length - spliceIndex, ...action.pointsFromClose)
+          self.points.splice(0, 0, ...action.pointsFromDrag)
+        } else {
+          let spliceIndex = (action.isReversed)
+            ? action.pointIndexClose
+            : action.pointIndexDrag
+
+          self.splicePoints = self.splicePoints.concat(self.points.slice(spliceIndex, (spliceIndex + action.pointsSpliceCount + 1)));
+
+          if (action.isReversed) {
+            self.splicePoints = self.splicePoints.reverse()
+          }
+
+          // we add 1 because we don't splice out either drag/close point
+          self.points.splice(spliceIndex + 1, action.pointsSpliceCount, ...action.pointsFromDrag)
+        }
+        self.spliceActive = true
+        self.spliceReverse = action.isReversed
+        self.spliceThroughBeginning = action.throughBeginning
+        self.spliceDragPointIndex = action.pointIndexDrag
+        self.spliceClosePointIndex = action.pointIndexClose
+
+        self.setDragPoint(self.splicePoints.at(-1))
+        self.setClosePoint(self.points[self.spliceClosePointIndex])
+
+        self.drawActions.pop()
+      }
+
+      self.redoActions.push(action)
     },
 
-    setClipPath(points = []) {
-      self.clipPath = points
+    redo() {
+      if (self.redoActions.length == 0) return
+
+      const action = toJS(self.redoActions.at(-1))
```

Comment:
```suggestion
      const action = self.redoActions.at(-1)
```

`toJS` can be removed here too.